supersuit/generic_wrappers/frame_stack.py

```diff
@@ -23,19 +23,70 @@ def modify_obs_space(self, obs_space):
                 )
 
             self.old_obs_space = obs_space
-            self.observation_space = stack_obs_space(obs_space, stack_size)
+            self.observation_space = stack_obs_space(obs_space, stack_size, stack_dim0)
             return self.observation_space
 
         def reset(self, seed=None):
-            self.stack = stack_init(self.old_obs_space, stack_size)
+            self.stack = stack_init(self.old_obs_space, stack_size, stack_dim0)
+            self.reset_flag = True
 
         def modify_obs(self, obs):
             self.stack = stack_obs(
                 self.stack,
                 obs,
                 self.old_obs_space,
                 stack_size,
+                stack_dim0
             )
+
+            return self.stack
+
+        def get_last_obs(self):
+            return self.stack
+
+    return shared_wrapper(env, FrameStackModifier)
+
+
+def frame_stack_v2(env, stack_size=4, stack_dim0=False):
+    assert isinstance(stack_size, int), "stack size of frame_stack must be an int"
+
+    class FrameStackModifier(BaseModifier):
+        def modify_obs_space(self, obs_space):
+            if isinstance(obs_space, Box):
+                assert 1 <= len(obs_space.shape) <= 3, "frame_stack only works for 1, 2 or 3 dimensional observations"
+            elif isinstance(obs_space, Discrete):
+                pass
+            else:
+                assert False, "Stacking is currently only allowed for Box and Discrete observation spaces. The given observation space is {}".format(obs_space)
+
+            self.old_obs_space = obs_space
+            self.observation_space = stack_obs_space(obs_space, stack_size, stack_dim0)
+            return self.observation_space
+
+        def reset(self, seed=None):
+            self.stack = stack_init(self.old_obs_space, stack_size, stack_dim0)
+            self.reset_flag = True
+
+        def modify_obs(self, obs):
+            if self.reset_flag:
+                for _ in range(stack_size):
+                    self.stack = stack_obs(
+                        self.stack,
+                        obs,
+                        self.old_obs_space,
+                        stack_size,
+                        stack_dim0
+                    )
+                self.reset_flag = False
```

Comment:
Is a for loop necessary here? Is it not possible to simply do something like `tile_shape * obs`? This is just to make things run a little bit faster.

I'm ok if a completely new function was created just for this functionality as well.

Reply:
Your right. Actually, this is how frame_stack_v0 was implemented before, so I left it as it is. If you want, we can modify it, but I think it makes the code (very x 3) little faster 😅

Reply:
If it makes the code a lot less readable then we can just keep the original implementation, but if it's simple to implement, I think changing it would be better. Small optimizations like this add up a lot in the long run (especially for python). :smile: